Sources/Utility/Process.swift

```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    internal func validateArgumentIsExecutable(_ program: String) throws {
+
+        // check to see if the argument is already listed in the vaildated cache
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
+        guard let executableLocation = Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) else {
+            throw Process.Error.missingProgram(program: program)
+        }
+        guard localFileSystem.isExecutableFile(executableLocation) else {
+            throw Process.Error.programNotExecutable(program: program)
+        }
+        // add the program name to the validated cache
+        Process.validatedExecutables.insert(program)
```

Comment:
By store the result I meant, store the result of the lookup query. Since we don't need the actual path, we can store the result as a bool where true means we found it, false means we did not find it.

For now, lets leave the cache and it can be an add on later. I'll point you to examples in SwiftPM for thread safety.